### PR TITLE
TS-4543: Fix core in HttpSM::tunnel_handler_post.

### PR DESCRIPTION
**Pull request by shinrich**

### CODE REVIEW

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/286/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/178/ for details.

File: proxy/http/HttpSM.cc [hunk 1]
```diff
@@ -2841,9 +2844,9 @@ HttpSM::tunnel_handler_push(int event, void *data)
   ink_assert(data == &tunnel);
 
   // Check to see if the client is still around
-  HttpTunnelProducer *ua = tunnel.get_producer(ua_session);
+  HttpTunnelProducer *ua = (ua_session) ? tunnel.get_producer(ua_session) : tunnel.get_producer(HT_HTTP_CLIENT);
 
-  if (!ua->read_success) {
+  if (ua && !ua->read_success) {
```
Comment (oknet) [inline proxy/http/HttpSM.cc]: compare to the modify in tunnel_handler_post, can we use "if (!ua || !ua->read_success) {" here.

File: proxy/http/HttpSM.cc [hunk 2]
```diff
@@ -2722,7 +2722,10 @@ HttpSM::tunnel_handler_post(int event, void *data)
 {
   STATE_ENTER(&HttpSM::tunnel_handler_post, event);
 
-  HttpTunnelProducer *p = tunnel.get_producer(ua_session);
+  HttpTunnelProducer *p = ua_session != NULL ? tunnel.get_producer(ua_session) : tunnel.get_producer(HT_HTTP_CLIENT);
+  if (!p)
+    return 0; // Cannot do anything if there is no producer
```
Comment (oknet) [inline proxy/http/HttpSM.cc]: Do we have to set terminate_sm = true before return 0 ?

**Pull request closed without merging**